akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -364,21 +364,57 @@ object ByteString {
   }
 
   /**
+   * This class helps to optimise the access time for linear access such as looping all items of [[ByteStrings]].
+   *
+   * @param lastIndex the index of the last access
+   * @param arrayIndex the array index of the last access
+   * @param seen the amount of elements being seen, excluding the elements in current array
+   */
+  private final case class LinearScanHint(lastIndex: Int, arrayIndex: Int, seen: Int)
+
+  /**
    * A ByteString with 2 or more fragments.
    */
   final class ByteStrings private (private[akka] val bytestrings: Vector[ByteString1], val length: Int) extends ByteString with Serializable {
     if (bytestrings.isEmpty) throw new IllegalArgumentException("bytestrings must not be empty")
     if (bytestrings.head.isEmpty) throw new IllegalArgumentException("bytestrings.head must not be empty")
 
+    /**
+     * Gives a hint what the last access state is.
+     * This will optimise the access to constant time when the hint is correct,
+     * however degrades to linear time if the hint is wrong.
+     */
+    private var _linearScanHint = LinearScanHint(-1, -1, 0)
```

Comment:
the hint should *not* be in ByteStrings, it must be in a view – very much like an iterator but specialized